crates/roc_host/src/heap.rs

```diff
@@ -0,0 +1,244 @@
+
+impl<T> ThreadSafeRefcountedResourceHeap<T> {
+    pub fn new(max_elements: usize) -> Result<ThreadSafeRefcountedResourceHeap<T>> {
+        RefcountedResourceHeap::new(max_elements).map(|heap| ThreadSafeRefcountedResourceHeap {
+            heap: UnsafeCell::new(heap),
+            guard: Mutex::new(()),
+        })
+    }
+
+    pub fn alloc_for(self: &Self, data: T) -> Result<RocBox<()>> {
+        let _g = self.guard.lock().unwrap();
+        unsafe { &mut *self.heap.get() }.alloc_for(data)
+    }
+
+    pub fn dealloc<U>(self: &Self, ptr: *const U) {
+        let _g = self.guard.lock().unwrap();
+        unsafe { &mut *self.heap.get() }.dealloc(ptr)
+    }
+
+    // This is safe to call at any time with no lock!
+    pub fn in_range<U>(self: &Self, ptr: *const U) -> bool {
+        unsafe { &*self.heap.get() }.in_range(ptr)
+    }
+
+    pub fn box_to_resource<'a>(data: RocBox<()>) -> &'a mut T {
+        RefcountedResourceHeap::box_to_resource(data)
+    }
+}
+
+unsafe impl<T> Sync for ThreadSafeRefcountedResourceHeap<T> {}
+unsafe impl<T> Send for ThreadSafeRefcountedResourceHeap<T> {}
+
+#[repr(C)]
+struct Refcounted<T>(usize, T);
+
+/// HeapOfRefcounted is a wrapper around Heap for data that roc stores with a refcount.
+/// It will return a pointer to right after the refcount (what a `Box {}` would expect in Roc).
+pub struct RefcountedResourceHeap<T>(Heap<Refcounted<T>>);
+
+impl<T> RefcountedResourceHeap<T> {
+    pub fn new(max_elements: usize) -> Result<RefcountedResourceHeap<T>> {
+        Heap::new(max_elements).map(|heap| RefcountedResourceHeap(heap))
+    }
+
+    pub fn alloc_for(self: &mut Self, data: T) -> Result<RocBox<()>> {
+        self.0.alloc().map(|alloc_ptr| {
+            unsafe { *alloc_ptr = Refcounted(REFCOUNT_ONE, data) };
+            let box_ptr = alloc_ptr as usize + mem::size_of::<usize>();
+            unsafe { std::mem::transmute(box_ptr) }
+        })
+    }
+
+    pub fn dealloc<U>(self: &mut Self, ptr: *const U) {
+        self.0.dealloc(ptr as _);
+    }
+
+    pub fn in_range<U>(self: &Self, ptr: *const U) -> bool {
+        self.0.in_range(ptr as _)
+    }
+
+    pub fn box_to_resource<'a>(data: RocBox<()>) -> &'a mut T {
+        let box_ptr: usize = unsafe { std::mem::transmute(data) };
+
+        let alloc_ptr = (box_ptr - mem::size_of::<usize>()) as *mut Refcounted<T>;
+        let alloc: &mut Refcounted<T> = unsafe { &mut *alloc_ptr };
+        &mut alloc.1
+    }
+}
+
+/// The Heap is one mmap of data that can be interpreted multiple ways.
+///
+/// It can be view as list of unions between `T` and `usize`.
+/// In the case of a `T`, it is an allocated element.
+/// In the case of a `usize`, it is part of the freed list.
+/// The value of the `usize` is the next free node.
+///
+/// To Decide: This impl is really unsafe.
+/// The free list will get messed up if someone double frees.
+/// It has no checks in general.
+/// Should we add some sort of tagging or chunk info to protect against this?
```

Comment:
I don't think that's necessary, since the only possibility of double-free would come from the host itself, right?

Reply:
Or a bug in roc refcounting

Reply:
Gotcha, yeah in general I think both hosts and the Roc compiler backends should go for performance over guarding against memory safety problems with runtime checks, whereas platform APIs and applications should take the opposite approach.

In other words, I think application authors should assume memory safety has been taken care of and not need to worry about it (and platform APIs should preserve that invariant), whereas host authors and compiler backend authors should focus on performance over runtime checks on memory safety. That way application authors can get the "don't worry about it" memory safety experience while also having programs that run as fast as possible, given that constraint!